src/frontend/Deprecation_analysis.ml

```diff
@@ -0,0 +1,168 @@
+open Core_kernel
+open Ast
+open Middle
+
+let deprecated_functions =
+  String.Map.of_alist_exn
+    [ ("multiply_log", "lmultiply")
+    ; ("binomial_coefficient_log", "lchoose")
+    ; ("integrate_ode", "integrate_ode_rk45")
+    ; ("cov_exp_quad", "gp_cov_exp_quad") ]
+
+let deprecated_distributions =
+  String.Map.of_alist_exn
+    (List.concat_map Middle.Stan_math_signatures.distributions
+       ~f:(fun (fnkinds, name, _) ->
+         List.filter_map fnkinds ~f:(function
+           | Lpdf -> Some (name ^ "_log", name ^ "_lpdf")
+           | Lpmf -> Some (name ^ "_log", name ^ "_lpmf")
+           | Cdf -> Some (name ^ "_cdf_log", name ^ "_lcdf")
+           | Ccdf -> Some (name ^ "_ccdf_log", name ^ "_lccdf")
+           | Rng | UnaryVectorized -> None ) ))
+
+let deprecated_userdefined : string String.Table.t = String.Table.create ()
+
+let is_distribution name =
+  Option.is_some (String.Map.find deprecated_distributions name)
+
+let rename_distribution name =
+  Option.value ~default:name (String.Map.find deprecated_distributions name)
+
+let rename_function name =
+  Option.value ~default:name (String.Map.find deprecated_functions name)
+
+let distribution_suffix name =
+  String.is_suffix ~suffix:"_lpdf" name
+  || String.is_suffix ~suffix:"_lpmf" name
+  || String.is_suffix ~suffix:"_lcdf" name
+  || String.is_suffix ~suffix:"_lccdf" name
+
+let userdef_distributions stmts =
+  List.filter_map
+    ~f:(function
+      | {stmt= FunDef {funname= {name; _}; _}; _} ->
+          if
+            String.is_suffix ~suffix:"_log_lpdf" name
```

Comment:
If you feel like doing some refactoring while you're in there (I know you didn't originate this code) there are a couple of ways to avoid typing String so many times: 
* `let open String in` at the top, or
* wrap the expression with `String.()`

Reply:
Thanks, changed. I currently use `let open String in` in 4 different functions - maybe it's better to open in the global scope (hope I'm getting the terminology right)?

Reply:
Hm. Style-wise OCaml seems to prefer few global opens, but this module does revolve around strings quite a bit, heh. If one were to import String it would sort of be equivalent to saying this module's main type that it deals with is strings though, and I think maybe that's not quite true. The reason it's that strong is that String contains many things besides the functions we're using, like a type `t` and possibly other things that would be confusing and maybe even collide. So I'd say just keep the local opens or specifically import functions using `let is_suffix = String.is_suffix` :/

Reply:
OK, keeping the local opens than.